### PR TITLE
[SPARK-15181][ML][PYSPARK] Python API for GLR summaries.

### PR DESCRIPTION
**Pull request by sethah**
## What changes were proposed in this pull request?

This patch adds a python API for generalized linear regression summaries (training and test). This helps provide feature parity for Python GLMs.
## How was this patch tested?

Added a unit test to `pyspark.ml.tests`

### CODE REVIEW

**Comment by sethah** [comment]
cc @yanboliang @jkbradley This is part of Python API audit for Spark 2.0 QA. Could you take a look?

File: python/pyspark/ml/regression.py [hunk 1]
```diff
@@ -1337,6 +1338,204 @@ def intercept(self):
...
+    @since("2.0.0")
+    def predictions(self):
+        """
+        Predictions output by the model's `transform` method.
+        """
+        return self._call_java("predictions")
+
+    @property
+    @since("2.0.0")
+    def predictionCol(self):
```
Comment (holdenk) [inline python/pyspark/ml/regression.py]: Maybe consider copying the Scaladoc here
`Field in "predictions" which gives the prediction value of each instance. This is set to a new column name if the original model's predictionCol is not set.`?

Comment (sethah) [inline python/pyspark/ml/regression.py]: Done, thanks!

Comment (MLnick) [inline python/pyspark/ml/regression.py]: @sethah you changed this from the Scala ".. gives the prediction value of ..." to use "predicted", which I prefer. Do you mind updating the Scala doc to match?

File: python/pyspark/ml/tests.py [hunk 2]
```diff
@@ -905,6 +906,43 @@ def test_linear_regression_summary(self):
         sameSummary = model.evaluate(df)
         self.assertAlmostEqual(sameSummary.explainedVariance, s.explainedVariance)
 
+    def test_linear_regression_summary(self):
```
Comment (MLnick) [inline python/pyspark/ml/tests.py]: this method is the same name as the one above - should it perhaps be `test_glm_summary`?

**Comment by MLnick** [comment]
A couple minor comments, pending that LGTM

File: python/pyspark/ml/regression.py (hunk 1 above)
Comment (sethah) [inline python/pyspark/ml/regression.py]: Done.

File: python/pyspark/ml/tests.py (hunk 2 above)
Comment (sethah) [inline python/pyspark/ml/tests.py]: Yep, thanks!

**Comment by MLnick** [comment]
Will leave open for a day in case @yanboliang or @jkbradley want to take a pass.

File: python/pyspark/ml/regression.py [hunk 3]
```diff
@@ -1337,6 +1338,204 @@ def intercept(self):
...
+    @property
+    @since("2.0.0")
+    def summary(self):
+        """
+        Gets summary (e.g. residuals, deviance, pValues) of model on
+        training set. An exception is thrown if
+        `trainingSummary is None`.
+        """
+        java_lrt_summary = self._call_java("summary")
+        return GeneralizedLinearRegressionTrainingSummary(java_lrt_summary)
```
Comment (yanboliang) [inline python/pyspark/ml/regression.py]: Minor: ```java_lrt_summary``` -> ```java_grt_summary```. I guess you refer the code of ```LinearRegressionModel``` in regression.py.

File: python/pyspark/ml/regression.py [hunk 4]
```diff
@@ -1337,6 +1338,204 @@ def intercept(self):
...
+    def residualDegreeOfFreedomNull(self):
+        """
+        The residual degrees of freedom for the null model.
+        """
+        return self._call_java("residualDegreeOfFreedomNull")
+
+    @since("2.0.0")
+    def residuals(self, residualsType="deviance"):
+        """
+        Get the residuals of the fitted model by type.
```
Comment (yanboliang) [inline python/pyspark/ml/regression.py]: Minor: Add doc for default returned residuals (deviance residuals).

**Comment by yanboliang** [comment]
LGTM except the two minor issues.

**Comment by sethah** [comment]
@yanboliang thanks for the review! I addressed your comments.

**Comment by holdenk** [comment]
You probably want to merge in master the PySpark tests have been updated to use `SparkSession`

**Comment by MLnick** [comment]
Merged to master and branch-2.0. Thanks!

**Pull request closed without merging**